### ISSUE TITLE
03: n-gram language model end-to-end through the shared eval harness

### ISSUE THREAD

**Issue by JasonOng0109** [JasonOng0109/nlp#3]
## Parent
#1

## What to build
The shared evaluation harness and the n-gram baseline together, so the harness is proven by a real model before any training loop exists. Perplexity is computed once, in one place, over an explicit list of target positions that every future model will reuse.

## Acceptance criteria
- [ ] An interpolated n-gram over orders 1 to 5 with add-k smoothing is fitted on `lm_train`
- [ ] Interpolation weights are tuned on `lm_valid`; the tuning procedure and the selected weights are recorded in the metrics artifact
- [ ] A test asserts every interpolated distribution sums to one, at every order
- [ ] The evaluation harness exposes a target-position list and scores a model against it; a test asserts two different callers receive an identical list for the same protocol
- [ ] `lm_test` perplexity and average next-token cross-entropy are emitted to a metrics artifact
- [ ] The metrics artifact records the full argument set of the run
- [ ] The stage runs on the fixture corpus in under a minute

## Blocked by
- #2

**Comment by JasonOng0109** [comment]
Implemented in 4c18c66 and 3841fb2 on `main`. 85 tests pass. **`lm_test` perplexity 141.40**, cross-entropy 4.9516 nats over 672,011 targets.

`python -m scripts.fit_ngram --source imdb [--refit]`

**Harness.** The harness owns the target list and the arithmetic, so no model can choose the positions it is scored on. `headline_positions` is the protocol ADR-0001 names as separate from the context sweep: every `lm_test` target carrying at least `MIN_CONTEXT` tokens of history, which costs four targets out of 672,015 and buys an identical target list for every model. A uniform model over V scores perplexity V, which is the one closed form available to check the harness's own arithmetic against.

The model interface is a single `log_probs(stream, positions, context_length)`. The cap is what #7 will sweep; the n-gram honours it by dropping to effective order `context_length + 1`, which is how a count-based model lands on the same axis as the recurrent and attention models. Already tested.

**Counts.** Sorted arrays with binary-search lookup, not dictionaries: the real corpus holds 27.9M distinct n-grams over orders 1–5, about 3.1GB as a dict of tuples against 0.67GB here. Keys are big-endian so byte order is numeric order and n-grams sharing a prefix stay contiguous. Context counts are built only from contexts that are followed by a token, which is what makes add-k sum to one exactly rather than nearly; a test asserts it at every order.

**Tuning.** EM for the weights, a grid for `k`, both on `lm_valid`. EM rather than a grid for the weights because five weights make a four-dimensional simplex no grid covers honestly.

Two methodology bugs the first real run had, now fixed and guarded by tests. The selected `k` sat at the edge of a grid of {1e-3 … 1}, so it had not been shown to be a minimum, and EM was hitting a 100-iteration cap without converging. Widening the grid to {1e-7 … 1} and raising the cap moved `lm_test` perplexity from 154.01 to **141.40**. The artifact now records `selected_k_at_grid_edge` and `all_trials_converged`.

| k | 1e-7 | **1e-6** | 1e-5 | 1e-4 | 1e-3 | 1e-2 | 0.1 | 1.0 |
|---|---|---|---|---|---|---|---|---|
| `lm_valid` ppl | 138.33 | **138.31** | 138.35 | 140.57 | 151.02 | 179.05 | 244.98 | 373.35 |

**Result.** Weights `[0.093, 0.499, 0.278, 0.087, 0.043]` for orders 1–5 — bigram and trigram carry 78% of the mixture. The artifact records why alongside: of `lm_test` targets, the order-5 n-gram was seen in training only 12.4% of the time, order 4 30.5%, order 3 62.9%, order 2 92.3%. EM put the mass where the evidence is.

**Add-k is near-inert, and is described as such.** `k` tunes to 1e-6, injecting `k * V = 0.02` pseudo-counts per context against a 20,003-word vocabulary, so this is in practice Jelinek-Mercer interpolation of maximum-likelihood estimates, with add-k surviving only to keep unseen n-grams positive. That is a property of add-k at a large vocabulary, not a tuning accident. ADR-0006 records the finding and why interpolated Kneser-Ney was declined; the artifact records `add_k_pseudo_counts_per_context` so the claim is checkable.

**Checkpointing.** The fitted tables are saved to `checkpoints/` and reused unless `--refit` is passed, since #7, #8 and #9 all need this model. A rerun drops from 318s to 63s and reproduces 141.40 exactly; a test asserts a second run reuses and reproduces. Peak memory 1.8GB, checkpoint 995MB.
